plat/renesas/rcar/aarch64/plat_helpers.S

```diff
@@ -0,0 +1,397 @@
+
+	/* ---------------------------------------------
+	 * plat_secondary_reset
+	 *
+	 * ---------------------------------------------
+	 */
+func plat_secondary_reset
+	mrs	x0, sctlr_el3
+	bic	x0, x0, #SCTLR_EE_BIT
+	msr	sctlr_el3, x0
+	isb
+
+	mrs	x0, cptr_el3
+	bic	w0, w0, #TCPAC_BIT
+	bic	w0, w0, #TTA_BIT
+	bic	w0, w0, #TFP_BIT
+	msr	cptr_el3, x0
+
+	mov_imm	x0, PARAMS_BASE
+	mov_imm	x2, BL31_BASE
+       ldr x3, =BOOT_KIND_BASE
+	mov x1, #0x1
+	str x1, [x3]
+	br	x2	/* jump to BL31 */
+	nop
+	nop
+	nop
+endfunc plat_secondary_reset
+
+	/* ---------------------------------------------
+	 * plat_enter_bl31
+	 *
+	 * ---------------------------------------------
+	 */
+func bl2_enter_bl31
+	mov	x20, x0
+        /*
+         * MMU needs to be disabled because both BL2 and BL31 execute
+         * in EL3, and therefore share the same address space.
+         * BL31 will initialize the address space according to its
+         * own requirement.
+         */
+	bl	disable_mmu_icache_el3
+	tlbi	alle3
+	ldp	x0, x1, [x20, #ENTRY_POINT_INFO_PC_OFFSET]
+	msr	elr_el3, x0
+	msr	spsr_el3, x1
+	eret
+endfunc bl2_enter_bl31
+
+	/* -----------------------------------------------------
+	 * void platform_mem_init (void);
+	 *
+	 * Zero out the mailbox registers in the shared memory
+	 * and set the rcar_boot_kind_flag.
+	 * The mmu is turned off right now and only the primary can
+	 * ever execute this code. Secondaries will read the
+	 * mailboxes using SO accesses.
+	 * -----------------------------------------------------
+	 */
+func platform_mem_init
+#if !IMAGE_BL2
+	ldr	x0, =MBOX_BASE
+	mov	w1, #PLATFORM_CORE_COUNT
+loop:
+	str	xzr, [x0], #CACHE_WRITEBACK_GRANULE
+	subs	w1, w1, #1
+	b.gt	loop
+	/* for R-Car Gen3		*/
+	ldr	x0, =rcar_boot_kind_flag
+	str	w21, [x0]
+#endif
+	ret
+endfunc platform_mem_init
+
+	/* ---------------------------------------------
+	 * void plat_report_exception(unsigned int type)
+	 * Function to report an unhandled exception
+	 * with platform-specific means.
+	 * ---------------------------------------------
+	 */
+func plat_report_exception
+#if IMAGE_BL2
+	mov	w1, #FIQ_SP_ELX
+	cmp	w0, w1
+	beq	rep_exec_fiq_elx
+	b	rep_exec_panic_type
+rep_exec_fiq_elx:
+	bl	gicv2_acknowledge_interrupt
+	mov	x2, #INT_ID_MASK
+	and	x0, x0, x2
+	mov	x1, #ARM_IRQ_SEC_WDT
+	cmp	x0, x1
+	bne	rep_exec_panic_id
+	mrs	x0, ELR_EL1
+	b	rcar_swdt_exec
+rep_exec_panic_type:	/* interrupt type error	*/
+	/* x0 is interrupt TYPE */
+	b	bl2_interrupt_error_type
+rep_exec_panic_id:	/* interrupt id error	*/
+	/* x0 is interrupt ID */
+	b	bl2_interrupt_error_id
+rep_exec_end:
+#endif
+	ret
+endfunc plat_report_exception
+
+	/* ---------------------------------------------
+	 * platform_is_primary_cpu
+	 *
+	 * ---------------------------------------------
+	 */
+func platform_is_primary_cpu
+	and	x0, x0, #MPIDR_CPU_MASK	/* Get the bit7-0 of MPIDR_EL1 -> x0	*/
+	cmp	x0, #RCAR_PRIMARY_CPU	/* x0 = 0 (primary CPU)?		*/
+	cset	x0, eq			/* If x0 is 0 carry bit set		*/
```

Comment:
See **Boot** section of **H3/M3 Device Manual** regarding selection of a master boot processor through the MD7 and MD6 pins.
R-Car H3 Salvator-SX board supports selection between A57/A53 boot CPU using the board's DIP switch and in case of A53 boot the primary CPU #0 will have MPIDR=0x80000100.
Note that this switch is not present on H3/M3 Starter Kit boards, and boot CPU is A57 #0 with MPIDR=0x80000000.
For correct detection of primary CPU, please consider the following modifications:
1) in plat\renesas\rcar\include\rcar_def.h add:
`/* Cluster IDs read from MPIDR */`
`#define RCAR_CLUSTER_A57		0`
`#define RCAR_CLUSTER_A53		1`
2) Modify platform_is_primary_cpu function as:
`func platform_is_primary_cpu`
`    ldr	x1, =RCAR_MODEMR			/* Mode pin register */`
`    and	x0, x0, #(MPIDR_CLUSTER_MASK | MPIDR_CPU_MASK)`
`    ldr	x1, [x1]				/* Read Mode pin register */`
`    and	x1, x1, #MODEMR_BOOT_CPU_MASK		/* Detect master boot processor in MD[7:6] */`
`    mov	x2, #RCAR_CLUSTER_A53			/* x2 = A53 cluster #1 */`
`    cmp	x1, #MODEMR_BOOT_CPU_CA57		/* A57 boot cluster?   */`
`    mov	x1, #RCAR_CLUSTER_A57			/* x1 = A57 cluster #0 */`
`    csel	x1, x1, x2, eq				/* Select boot cluster */` 		
`    cmp	x0, x1, lsl #MPIDR_AFFINITY_BITS	/* Master boot processor #0 ? */`
`    cset	x0, eq					/* If true, x0 = 1, otherwise 0 */`
`    ret`
`endfunc platform_is_primary_cpu`

Reply:
yes I have seen this but wasnt this code removed from earlier Renesas's versions? did you find out why?

Reply:
>  It's also interesting to note that the old code tries to detect if boot CPU is ARM R7 by execution AArch64 code.

That's certainly interesting...

Reply:
@AlexeiFedorov um but whoever replaced that function for its current implementation made a conscious decision (I dont think it was a mistake). The current code 'somehow' works as expected so I still think we should merge the implementation coming from Renesas until we have further clarification or we have a bug report (as I said I cant make it fail on my Salvator-x, can you?)

Reply:
@ldts. I'll have a look to work out why this code doesn't fail.

Reply:
awesome. I am working on the documentation and instructions. this is the tree that I will use to update the pull request tomorrow after I finished with the docs. https://github.com/ldts/arm-trusted-firmware (branch rcar_gen3/devel)

Reply:
@ldts 
Is it "branch **rcar_gen3/devel**" but not  **rcar_gen3/upstream**?